File: javascript/src/museo/catalogoObras.js
import { hoyUTC } from "./fechasUtil.js";

export class CatalogoObras {
  constructor() {
    /** @type {import('./obraArte.js').ObraArte[]} */
    this._obras = [];
    this.fechaActualizacion = hoyUTC();
  }

  tocar() {
    this.fechaActualizacion = hoyUTC();
  }

  getFechaActualizacion() {
    return this.fechaActualizacion;
  }

  listarObras() {
    return [...this._obras];
  }

  buscarObra(idObra) {
    for (const o of this._obras) {
      if (o.getIdObra() === idObra) {
        return o;
      }
    }
    throw new Error(`No existe obra con id ${idObra}`);
  }

  agregarObra(obra) {
    this._obras.push(obra);
    this.tocar();
  }

  eliminarObra(idObra) {
    this._obras = this._obras.filter((o) => o.getIdObra() !== idObra);
    this.tocar();
  }

  contieneObra(obra) {
    return this._obras.includes(obra);
  }
}

File: java/src/museo/CatalogoObras.java
package museo;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.NoSuchElementException;

public class CatalogoObras {

    private final ArrayList<ObraArte> obras = new ArrayList<>();
    private LocalDate fechaActualizacion = LocalDate.now();

    void tocar() {
        this.fechaActualizacion = LocalDate.now();
    }

    public LocalDate getFechaActualizacion() {
        return fechaActualizacion;
    }

    public List<ObraArte> listarObras() {
        return Collections.unmodifiableList(new ArrayList<>(obras));
    }

    public ObraArte buscarObra(int idObra) {
        for (ObraArte o : obras) {
            if (o.getIdObra() == idObra) {
                return o;
            }
        }
        throw new NoSuchElementException("No existe obra con id " + idObra);
    }

    public void agregarObra(ObraArte obra) {
        obras.add(obra);
        tocar();
    }

    public void eliminarObra(int idObra) {
        obras.removeIf(o -> o.getIdObra() == idObra);
        tocar();
    }

    public boolean contieneObra(ObraArte obra) {
        return obras.contains(obra);
    }
}
